jupyterhubutils/authenticator/lsstgithubauth.py

```diff
@@ -0,0 +1,170 @@
+    def _check_github_denylist(self, userdict, denylist):
+        if ("auth_state" not in userdict or not userdict["auth_state"]):
+            self.log.warning("User doesn't have auth_state: rejecting.")
+            return True
+        ast = userdict["auth_state"]
+        if ("access_token" not in ast or not ast["access_token"]):
+            self.log.warning("User doesn't have access token: rejecting.")
+            return True
+        tok = ast["access_token"]
+        gh_org = yield self._get_github_user_organizations(tok)
+        if not gh_org:
+            self.log.warning("Could not get list of GH user orgs: rejecting.")
+            return True
+        deny = list(set(gh_org) & set(denylist))
+        if deny:
+            self.log.warning("User in denylist %s: rejecting." % str(deny))
+            return True
+        self.log.debug("User not in denylist %s" % str(denylist))
+        return False
+
+    @gen.coroutine
+    def _get_github_user_organizations(self, access_token):
+        # Requires 'read:org' token scope.
+        http_client = AsyncHTTPClient()
+        headers = github_api_headers(access_token)
+        gh_api = self.lsst_mgr.config.github_api
+        next_page = "https://{}/user/orgs".format(gh_api)
+        orgmap = {}
+        while next_page:
+            req = HTTPRequest(next_page, method="GET", headers=headers)
+            try:
+                resp = yield http_client.fetch(req)
+            except HTTPError as exc:
+                self.log.error("{} -> {}".format(next_page, exc))
+                return None
+            resp_json = json.loads(resp.body.decode('utf8', 'replace'))
+            next_page = next_page_from_links(resp)
+            for entry in resp_json:
+                # This could result in non-unique groups, if the first 32
+                #  characters of the group names are the same.
+                normalized_group = entry["login"][:32]
+                orgmap[normalized_group] = entry["id"]
+        return orgmap
+
+    @gen.coroutine
+    def _get_github_user_email(self, access_token):
+        # Determine even private email, if the token has 'user:email'
+        #  scope
+        http_client = AsyncHTTPClient()
+        headers = github_api_headers(access_token)
+        gh_api = self.lsst_mgr.config.github_api
+        next_page = "https://{}/user/emails".format(gh_api)
+        while next_page:
+            req = HTTPRequest(next_page, method="GET", headers=headers)
+            resp = yield http_client.fetch(req)
+            resp_json = json.loads(resp.body.decode('utf8', 'replace'))
+            next_page = next_page_from_links(resp)
+            for entry in resp_json:
+                if "email" in entry:
+                    if "primary" in entry and entry["primary"]:
+                        return entry["email"]
+        return None
+
+    @gen.coroutine
+    def pre_spawn_start(self, user=None, spawner=None):
+        update_env = {}
+        # Github fields
```

Comment:
Maybe put a comment here about how this is to configure git inside of the lab pod rather than having anything to do with actually authenticating

Reply:
method removed